tasks/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+                        task.process_data(data)
+                    }) {
+                        return Ok(SchedulerEvent::Unsolicited(Response::Data(data)));
+                    }
+                }
+                ClientFlowEvent::StatusReceived { status } => match StatusV2::from(status) {
+                    StatusV2::Untagged(body) => {
+                        if let Some(body) =
+                            trickle_down(body, self.active_tasks_mut(), |task, body| {
+                                task.process_untagged(body)
+                            })
+                        {
+                            return Ok(SchedulerEvent::Unsolicited(Response::Status(
+                                Status::from(StatusV2::Untagged(body)),
+                            )));
+                        }
+                    }
+                    StatusV2::Bye(bye) => {
+                        if let Some(bye) =
+                            trickle_down(bye, self.active_tasks_mut(), |task, bye| {
+                                task.process_bye(bye)
+                            })
+                        {
+                            return Ok(SchedulerEvent::Unsolicited(Response::Status(
+                                Status::from(StatusV2::Bye(bye)),
+                            )));
+                        }
+                    }
+                    StatusV2::Tagged(Tagged { tag, body }) => {
+                        let handle = {
+                            if let Some((handle, _)) =
+                                self.active_tasks.iter().find(|(_, (tag_, _))| tag == *tag_)
+                            {
+                                *handle
+                            } else {
+                                return Err(SchedulerError::FatalProtocolIssue(
+                                    FatalProtocolIssue::UnexpectedTaggedResponse,
+                                ));
+                            }
+                        };
+
+                        let (_, task) = self.active_tasks.remove(&handle).unwrap();
+
+                        let output = Some(task.process_tagged(body));
+
+                        return Ok(SchedulerEvent::TaskFinished(TaskToken { handle, output }));
+                    }
+                },
+            }
+        }
+    }
+
+    fn active_tasks_mut(&mut self) -> impl Iterator<Item = &mut Box<dyn TaskAny>> {
+        self.active_tasks.values_mut().map(|(_, task)| task)
+    }
+}
+
+#[derive(Debug)]
+pub enum SchedulerEvent {
+    TaskFinished(TaskToken),
+    Unsolicited(Response<'static>),
+}
+
+#[derive(Debug, Error)]
+pub enum SchedulerError {
+    /// Flow error
+    #[error("flow error")]
+    Flow(#[from] ClientFlowError),
+    /// Fatal protocol issue
+    #[error("fatal protocol issue")]
+    FatalProtocolIssue(FatalProtocolIssue),
+}
+
+#[derive(Debug, Error)]
+pub enum FatalProtocolIssue {
```

Comment:
Do we really need this enum? Isn't `ClientFlowError` also a fatal protocol issue?

Reply:
Not sure... but let's err on the side of less code :+1: